Multi-target and accession-based mapping

sallybg opened the pull request.

sallybg committed:
Multi-target mapping partial draft: alignment and transcript selection

sallybg committed:
Make it the default behavior to prefer genomic mappings when available

sallybg committed:
Map score sets with multiple targets

sallybg committed:
fix: single-target blat result named incorrectly

sallybg committed:
Remove blank reference sequence entries from output

sallybg committed:
Fix type annotations for target gene metadata param

sallybg committed:
Accession-based mapping and add cdot to environment

sallybg committed:
API support for accession-based mapping

Note that this is mostly structured to handle multi-target mapping,
but does not contain all changes required for multi-target mapping
(specifically final output / reference sequence structure).
Such changes would require corresponding changes in mavedb-api
which we are not prepared to deploy yet.

sallybg committed:
Temporarily remove support for multi-target score sets

sallybg committed:
Corrections for accession-based mapping without multi-target mapping

sallybg committed:
Bug fixes for temporary non-multi-target staging deployment

sallybg committed:
Update UTA DB version

sallybg committed:
Re-implement multi-target mapping

sallybg committed:
Add transcript information to mapped metadata for genomic score sets

sallybg committed:
Fix CLI prefer_genomic flag

bencap commented on src/api/routers/map.py:
```diff
@@ -37,13 +43,11 @@ async def map_scoreset(urn: str, store_path: Path | None = None) -> ScoresetMapp
     """Perform end-to-end mapping for a scoreset.
```
I noticed this function is a little inconsistent with its return types. Sometimes it returns a ScoresetMapping, and other times it returns a JSONResponse containing a ScoresetMapping. We should probably make that consistent and choose one or the other.

bencap commented on src/api/routers/map.py:
```diff
@@ -75,77 +79,145 @@ async def map_scoreset(urn: str, store_path: Path | None = None) -> ScoresetMapp
         )
-    if vrs_results is None:
+    # TODO this should instead check if all values in dict are none. or might not need this at all.
+    if vrs_results is None or len(vrs_results) == 0:
```
Yeah, since `vrs_results` can never be None now (we assign an empty dict to it up front), this should probably be checking to make sure `vrs_results` has items in it and that at least one of those values is not None.
```suggestion
if not vrs_results or all(mapping_result is None for mapping_result in vrs_results.values()):
```

bencap commented on src/api/routers/map.py:
```diff
@@ -75,77 +79,145 @@ async def map_scoreset(urn: str, store_path: Path | None = None) -> ScoresetMapp
         )
-    if vrs_results is None:
+    # TODO this should instead check if all values in dict are none. or might not need this at all.
+    if annotated_vrs_results is None or len(annotated_vrs_results) == 0:
```
Same as prior comment

bencap commented on src/dcd_mapping/align.py:
```diff
@@ -143,50 +146,77 @@ def _write_blat_output_tempfile(result: subprocess.CompletedProcess) -> str:
-            output = read_blat(out_file, "blat-psl")
+            output = parse_blat(out_file, "blat-psl")
+
+        # TODO reevaluate this code block - are there cases in mavedb where target sequence type is incorrectly supplied?
```
I would hope not :)

bencap replied:
Do you think it's possible any of the weird mapping outputs we have seen are coming from cases where we enter into this block? Or do you think we never enter this block with existing data.

bencap commented on src/dcd_mapping/align.py:
```diff
@@ -143,50 +146,77 @@ def _write_blat_output_tempfile(result: subprocess.CompletedProcess) -> str:
     :param scoreset_metadata: object containing scoreset attributes
     :param silent: suppress BLAT command output
-    :return: BLAT query result
+    :return: dict where keys are target gene identifiers and values are BLAT query result objects
```
We should still be able to type this return value with dict[str, QueryResult].

bencap commented on src/dcd_mapping/align.py:
```diff
@@ -291,12 +320,132 @@ def _get_best_match(output: QueryResult, metadata: ScoresetMetadata) -> Alignmen
+            target_label not in scoreset_metadata.target_genes
+            and len(scoreset_metadata.target_genes) == 1
+        ):
+            target_label = list(scoreset_metadata.target_genes.keys())[0]  # noqa: RUF015
```
Is it still the case here that this doesn't work for multi-target score sets?

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -39,10 +39,13 @@
     ScoreAnnotationWithLayer,
```
More of a general comment on this is that we should use Cis-Phased Blocks for VRS2.0 and then convert those to Haplotypes for VRS1.3. I'm fine if that is a task we add to the backlog and accomplish after merging this in though.

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -406,20 +417,44 @@ def annotate(
+        if metadata.target_accession_id.startswith(("NP", "ENSP")):
+            seq_type = TargetSequenceType.PROTEIN
+        # TODO full list of transcript and contig accession id prefixes
+        elif metadata.target_accession_id.startswith(("NM", "ENST", "NC")):
```
Do these lists still need to be expanded?

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -406,20 +417,44 @@ def annotate(
-            raise ValueError
+        if tx_output is None or isinstance(tx_output, TxSelectError):
+            # TODO catch this error - don't stop whole job for one failed target
+            # raise ValueError
```
Based on lines 654-664 this seems like it is the case now, so we can probably drop this TODO

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -435,35 +470,54 @@ def _get_computed_reference_sequence(
+            # TODO catch this error, don't fail whole job for one target
+            # msg = "ID could not be acquired from Seqrepo for transcript identifier"
+            # raise ValueError(msg)
+            return None
```
Based on lines 654-664 this seems like it is the case now, so we can probably drop this TODO

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -435,35 +470,54 @@ def _get_computed_reference_sequence(
+        # TODO catch this error, don't fail whole job for one target
         msg = "ID could not be acquired from Seqrepo for chromosome identifier"
         raise ValueError(msg)
+        # return None
```
What is the reason for this error being raised while the other errors return None?

bencap commented on src/dcd_mapping/annotate.py:
```diff
@@ -513,64 +567,105 @@ def write_scoreset_mapping_to_json(
+    # set preferred layers for each target, to allow a mix of coding and noncoding targets
+    # TODO maybe this should be reevaluated and we should only allow one preferred layer per score set,
+    # since I can't imagine an experimental assay where some variants are assayed as nucleotide variants
+    # and others are assayed as amino acid variants.
```
Lets add a backlog item for this re-evaluation if you still think it's worthwhile. For now, it seems fine to leave. I probably tend to agree this only needs to be set at the score set level.

bencap commented on src/dcd_mapping/main.py:
```diff
@@ -223,7 +233,8 @@ async def map_scoreset(
         return
-    if vrs_results is None:
+    # TODO this should be if all values in dict are none. or might not need this at all.
+    if vrs_results is None or len(vrs_results) == 0:
```
Since we set `vrs_results` to an empty dictionary it can never be None, so this check could just be
```suggestion
    if not vrs_results:
```

bencap commented on src/dcd_mapping/main.py:
```diff
@@ -238,20 +249,32 @@ async def map_scoreset(
+            _emit_info(f"Score set mapping output saved to: {final_output}.", silent)
+            return
+    # TODO this should be if all values in dict are none. or might not need this at all.
+    if annotated_vrs_results is None:
```
Same thing here
```suggestion
if not vrs_results:
```

bencap commented on src/dcd_mapping/main.py:
```diff
@@ -238,20 +249,32 @@ async def map_scoreset(
+                metadata.urn,
+                vrs_version,
+            )
+        except Exception as e:  # TODO create AnnotationError class and replace ValueErrors in annotation steps with AnnotationErrors
```
Lets add a backlog item for this

bencap commented on src/dcd_mapping/mavedb_data.py:
```diff
@@ -174,50 +185,95 @@ def get_scoreset_metadata(
+
+    for gene in metadata["targetGenes"]:
+        if not _metadata_response_is_human(metadata):
+            # TODO allow score sets with mix of human and non-human targets? This may not come up, but is doable with a little restructuring.
```
Backlog item

bencap commented on src/dcd_mapping/mavedb_data.py:
```diff
@@ -174,50 +185,95 @@ def get_scoreset_metadata(
+            else:
+                # Should we quit the whole mapping job if this comes up, or just skip this row and only quit if none contain hgvs_nt or hgvs_pro?
+                msg = f"Each score row in {metadata.urn} must contain hgvs_nt or hgvs_pro variant description "
+                raise ScoresetNotSupportedError(msg)
```
Good question here. I think the answer to this is that we want a guarantee of at least a pre-mapped variant for every variant in a mapping query. If we can't provide that, we must return some sort of error like we do at present.

bencap commented on src/dcd_mapping/mavedb_data.py:
```diff
@@ -174,50 +185,95 @@ def get_scoreset_metadata(
+                # TODO check assumption of no colon in hgvs unless reference sequence identifier present
+                prefix = (
+                    row["hgvs_pro"].split(":")[0] if ":" in row["hgvs_pro"] else None
+                )
```
I believe this assumption is correct.

bencap reviewed (approved):
Thanks for all the work on this Sally, I know it was a lot to fit these new features into how this was originally built. I left a decent amount of comments, but all of them are pretty minor and that mostly stems from these being a large set of changes. Although I didn't personally run mapping jobs while testing these, it seems to me like the capability to map all of our pillar project data sets is a really reasonable test set.

Lets try and translate the TODOs here into the dcd mapping backlog. I tried to comment on most of them but might have missed a few and/or skipped some duplicates. That backlog isn't super fleshed out yet so it'd be good to get some items in there.

I think you were planning on fixing the tests before merging, but feel free to merge these changes once you do.

sallybg commented:
Note to self - bump mapper version

sallybg committed:
Change tests and fixtures to reflect multi-target mapper changes

sallybg committed:
Always return a custom JSON response type from mapping api

This change also checks that the number of mappings for a score set is greater than 0 and returns an error if not.

sallybg committed:
CLI: Check that number of mappings for a score set is greater than 0, and return an error if not.

This change also removes a TODO for which there is now a backlog entry.

sallybg committed:
Remove TODO comments for which backlog entries have been created

sallybg committed:
Fix BLAT-incompatible target names in multi-target score sets

BLAT automatically removes certain characters from query names, including removing all characters after a space.
If the BLAT result name does not match any target genes in the score set, attempt to match based on BLAT's query name shortening patterns.
If multiple matches (could happen if labels are something like "Gene 1" and "Gene 2", in which case both would be shortened to "Gene"), raise an error.

sallybg committed:
Bump mapper version

sallybg committed:
Use https for mock requests

sallybg committed:
Mock cdot data provider for tests

Although cdot fetching is not currently directly used in any tests,
the cdot REST data provider is imported into dcd_mapping.vrs_map.
The cdot data provider uses a ChainedSeqFetcher, to which specific
fasta file paths are passed in dcd_mapping.lookup.
This results in a FileNotFoundError during tests, since these fasta
files are not available outside of the mapper Docker container.
This fix uses a fake fasta file to generate a ChainedSeqFetcher,
so this change does not support actual cdot transcript fetches for
future tests.
